pkg/controller/ec2/natgateway/controller.go

```diff
@@ -0,0 +1,210 @@
+	}
+
+	// in a successful response, there should be one and only one object
+	if len(response.NatGateways) != 1 {
+		return managed.ExternalObservation{}, errors.New(errNotSingleItem)
+	}
+
+	observed := response.NatGateways[0]
+
+	cr.Status.AtProvider = ec2.GenerateNATGatewayObservation(observed)
+
+	switch cr.Status.AtProvider.State {
+	case v1beta1.NatGatewayStatusPending:
+		cr.SetConditions(runtimev1alpha1.Unavailable())
+	case v1beta1.NatGatewayStatusFailed:
+		cr.SetConditions(runtimev1alpha1.Unavailable().WithMessage(aws.StringValue(observed.FailureMessage)))
+	case v1beta1.NatGatewayStatusAvailable:
+		cr.SetConditions(runtimev1alpha1.Available())
+	case v1beta1.NatGatewayStatusDeleting:
+		cr.SetConditions(runtimev1alpha1.Deleting())
+	case v1beta1.NatGatewayStatusDeleted:
+		return managed.ExternalObservation{
+			ResourceExists: false,
+		}, nil
+	}
+
+	return managed.ExternalObservation{
+		ResourceExists:   true,
+		ResourceUpToDate: v1beta1.CompareTags(cr.Spec.ForProvider.Tags, observed.Tags),
+	}, nil
+}
+
+func (e *external) Create(ctx context.Context, mgd resource.Managed) (managed.ExternalCreation, error) {
+	cr, ok := mgd.(*v1beta1.NATGateway)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errUnexpectedObject)
+	}
+
+	cr.Status.SetConditions(runtimev1alpha1.Creating())
+	if err := e.kube.Status().Update(ctx, cr); err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errStatusUpdate)
+	}
+
+	nat, err := e.client.CreateNatGatewayRequest(&awsec2.CreateNatGatewayInput{
+		AllocationId: cr.Spec.ForProvider.AllocationID,
+		SubnetId:     cr.Spec.ForProvider.SubnetID,
+		TagSpecifications: []awsec2.TagSpecification{
+			{
+				ResourceType: "natgateway",
+				Tags:         v1beta1.GenerateEC2Tags(cr.Spec.ForProvider.Tags),
+			},
+		},
+	}).Send(ctx)
+	if err != nil {
+		return managed.ExternalCreation{}, errors.Wrap(err, errCreate)
+	}
+
+	meta.SetExternalName(cr, aws.StringValue(nat.NatGateway.NatGatewayId))
+
+	return managed.ExternalCreation{}, errors.Wrap(e.kube.Update(ctx, cr), errSpecUpdate)
+}
+
+func (e *external) Update(ctx context.Context, mgd resource.Managed) (managed.ExternalUpdate, error) {
+	cr, ok := mgd.(*v1beta1.NATGateway)
+	if !ok {
+		return managed.ExternalUpdate{}, errors.New(errUnexpectedObject)
+	}
+
+	response, err := e.client.DescribeNatGatewaysRequest(&awsec2.DescribeNatGatewaysInput{
+		NatGatewayIds: []string{meta.GetExternalName(cr)},
+	}).Send(ctx)
+	if err != nil {
+		return managed.ExternalUpdate{}, errors.Wrap(resource.Ignore(ec2.IsNatGatewayNotFoundErr, err), errDescribe)
+	}
+
+	// in a successful response, there should be one and only one object
+	if len(response.NatGateways) != 1 {
+		return managed.ExternalUpdate{}, errors.New(errNotSingleItem)
+	}
+
+	observed := response.NatGateways[0]
+
+	// This implies additional tags have been manually added to the resource
+	if len(observed.Tags) > len(cr.Spec.ForProvider.Tags) {
```

Comment:
I think this doesn't cover the case where tag number is same but the values or keys have changed. I'd suggest something similar to `DiffTags` we have here https://github.com/crossplane/provider-aws/blob/124095d/pkg/clients/aws.go#L369

Reply:
hmm `DiffTags` takes a `map[string]string` as input. Is it okay to add in code that converts `[]Tag` and `[]ec2.Tag` to `map[string][string]`?

Reply:
I usually lean towards having a separate function that works with `[]ec2.Tag` type without depending on that other one. Since the logic is pretty simple, known and straight-forwad, it's more flexible to have its duplicate that works with the types you need. Then we can use that function in other `ec2` resources since they all seem to be using the same `ec2.Tag` struct.

Reply:
This is handled properly now